// elysia/dist/cookie.mjs
// src/cookie.ts
import { parse } from "cookie";

// src/utils.ts
import { Kind } from "@sinclair/typebox";
import { Value as Value2 } from "@sinclair/typebox/value";
import { TypeCompiler } from "@sinclair/typebox/compiler";

// src/handler.ts
import { serialize } from "cookie";

// src/error.ts
import { Value } from "@sinclair/typebox/value";
var env = typeof Bun !== "undefined" ? Bun.env : typeof process !== "undefined" ? process?.env : void 0;
var ERROR_CODE = Symbol("ElysiaErrorCode");
var ELYSIA_RESPONSE = Symbol("ElysiaResponse");
var isProduction = (env?.NODE_ENV ?? env?.ENV) === "production";
var InvalidCookieSignature = class extends Error {
  constructor(key, message) {
    super(message ?? `"${key}" has invalid cookie signature`);
    this.key = key;
    this.code = "INVALID_COOKIE_SIGNATURE";
    this.status = 400;
  }
};

// src/handler.ts
var hasHeaderShorthand = "toJSON" in new Headers();

// src/utils.ts
var isBun = typeof Bun !== "undefined";
var hasHash = isBun && typeof Bun.hash === "function";
var signCookie = async (val, secret) => {
  if (typeof val !== "string")
    throw new TypeError("Cookie value must be provided as a string.");
  if (secret === null)
    throw new TypeError("Secret key must be provided.");
  const encoder = new TextEncoder();
  const secretKey = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const hmacBuffer = await crypto.subtle.sign(
    "HMAC",
    secretKey,
    encoder.encode(val)
  );
  const hmacArray = Array.from(new Uint8Array(hmacBuffer));
  const digest = btoa(String.fromCharCode(...hmacArray));
  return `${val}.${digest.replace(/=+$/, "")}`;
};
var unsignCookie = async (input, secret) => {
  if (typeof input !== "string")
    throw new TypeError("Signed cookie string must be provided.");
  if (null === secret)
    throw new TypeError("Secret key must be provided.");
  const tentativeValue = input.slice(0, input.lastIndexOf("."));
  const expectedInput = await signCookie(tentativeValue, secret);
  return expectedInput === input ? tentativeValue : false;
};
var isNumericString = (message) => {
  if (message.length < 16)
    return message.trim().length !== 0 && !Number.isNaN(Number(message));
  if (message.length === 16) {
    const numVal = Number(message);
    if (numVal.toString() === message)
      return message.trim().length !== 0 && !Number.isNaN(numVal);
  }
  return false;
};

// src/cookie.ts
var Cookie = class {
  constructor(_value, property = {}) {
    this._value = _value;
    this.property = property;
  }
  get() {
    return this._value;
  }
  get value() {
    return this._value;
  }
  set value(value) {
    if (typeof value === "object") {
      if (JSON.stringify(this.value) === JSON.stringify(value))
        return;
    } else if (this.value === value)
      return;
    this._value = value;
    this.sync();
  }
  add(config) {
    const updated = Object.assign(
      this.property,
      typeof config === "function" ? config(Object.assign(this.property, this.value)) : config
    );
    if ("value" in updated) {
      this._value = updated.value;
      delete updated.value;
    }
    this.property = updated;
    return this.sync();
  }
  set(config) {
    const updated = typeof config === "function" ? config(Object.assign(this.property, this.value)) : config;
    if ("value" in updated) {
      this._value = updated.value;
      delete updated.value;
    }
    this.property = updated;
    return this.sync();
  }
  remove(options) {
    if (this.value === void 0)
      return;
    this.set({
      domain: options?.domain,
      expires: /* @__PURE__ */ new Date(0),
      maxAge: 0,
      path: options?.path,
      sameSite: options?.sameSite,
      secure: options?.secure,
      value: ""
    });
  }
  get domain() {
    return this.property.domain;
  }
  set domain(value) {
    if (this.property.domain === value)
      return;
    this.property.domain = value;
    this.sync();
  }
  get expires() {
    return this.property.expires;
  }
  set expires(value) {
    if (this.property.expires?.getTime() === value?.getTime())
      return;
    this.property.expires = value;
    this.sync();
  }
  get httpOnly() {
    return this.property.httpOnly;
  }
  set httpOnly(value) {
    if (this.property.domain === value)
      return;
    this.property.httpOnly = value;
    this.sync();
  }
  get maxAge() {
    return this.property.maxAge;
  }
  set maxAge(value) {
    if (this.property.maxAge === value)
      return;
    this.property.maxAge = value;
    this.sync();
  }
  get path() {
    return this.property.path;
  }
  set path(value) {
    if (this.property.path === value)
      return;
    this.property.path = value;
    this.sync();
  }
  get priority() {
    return this.property.priority;
  }
  set priority(value) {
    if (this.property.priority === value)
      return;
    this.property.priority = value;
    this.sync();
  }
  get sameSite() {
    return this.property.sameSite;
  }
  set sameSite(value) {
    if (this.property.sameSite === value)
      return;
    this.property.sameSite = value;
    this.sync();
  }
  get secure() {
    return this.property.secure;
  }
  set secure(value) {
    if (this.property.secure === value)
      return;
    this.property.secure = value;
    this.sync();
  }
  toString() {
    return typeof this.value === "object" ? JSON.stringify(this.value) : this.value?.toString() ?? "";
  }
  sync() {
    if (!this.name || !this.setter)
      return this;
    if (!this.setter.cookie)
      this.setter.cookie = {
        [this.name]: Object.assign(this.property, {
          value: this.toString()
        })
      };
    else
      this.setter.cookie[this.name] = Object.assign(this.property, {
        value: this.toString()
      });
    return this;
  }
};
var createCookieJar = (initial, set, properties) => new Proxy(initial, {
  get(target, key) {
    if (key in target)
      return target[key];
    const cookie = new Cookie(
      void 0,
      properties ? { ...properties } : void 0
    );
    cookie.setter = set;
    cookie.name = key;
    return cookie;
  },
  set(target, key, value) {
    if (!(value instanceof Cookie))
      return false;
    if (!set.cookie)
      set.cookie = {};
    value.setter = set;
    value.name = key;
    value.sync();
    target[key] = value;
    return true;
  }
});
var parseCookie = async (set, cookieString, {
  secret,
  sign,
  ...properties
} = {}) => {
  if (!cookieString)
    return createCookieJar({}, set, properties);
  const jar = {};
  const isStringKey = typeof secret === "string";
  if (sign && sign !== true && !Array.isArray(sign))
    sign = [sign];
  const cookieKeys = Object.keys(parse(cookieString));
  for (let i = 0; i < cookieKeys.length; i++) {
    const key = cookieKeys[i];
    let value = parse(cookieString)[key];
    if (sign === true || sign?.includes(key)) {
      if (!secret)
        throw new Error("No secret is provided to cookie plugin");
      if (isStringKey) {
        value = await unsignCookie(value, secret);
        if (value === false)
          throw new InvalidCookieSignature(key);
      } else {
        let fail = true;
        for (let i2 = 0; i2 < secret.length; i2++) {
          const temp = await unsignCookie(value, secret[i2]);
          if (temp !== false) {
            value = temp;
            fail = false;
            break;
          }
        }
        if (fail)
          throw new InvalidCookieSignature(key);
      }
    }
    if (value === void 0)
      continue;
    const start = value.charCodeAt(0);
    if (start === 123 || start === 91)
      try {
        const cookie2 = new Cookie(JSON.parse(value));
        cookie2.setter = set;
        cookie2.name = key;
        jar[key] = cookie2;
        continue;
      } catch {
      }
    if (isNumericString(value))
      value = +value;
    else if (value === "true")
      value = true;
    else if (value === "false")
      value = false;
    const cookie = new Cookie(value, properties);
    cookie.setter = set;
    cookie.name = key;
    jar[key] = cookie;
  }
  return createCookieJar(jar, set);
};
export {
  Cookie,
  createCookieJar,
  parseCookie
};
